pingserver_test.go

```diff
@@ -39,67 +39,62 @@ func pingFail(_ context.Context, _ *connect.Request[pingv1.PingRequest]) (*conne
 	return nil, connect.NewError(connect.CodeDataLoss, errors.New("Oh no"))
 }
 
-func cumSumHappy(
+func pingStreamOkay(
 	ctx context.Context,
-	stream *connect.BidiStream[pingv1.CumSumRequest, pingv1.CumSumResponse],
+	stream *connect.BidiStream[pingv1.PingStreamRequest, pingv1.PingStreamResponse],
 ) error {
 	for {
 		if err := ctx.Err(); err != nil {
 			return err
 		}
-		request, err := stream.Receive()
+		msg, err := stream.Receive()
 		if err != nil && errors.Is(err, io.EOF) {
 			return nil
 		} else if err != nil {
 			return fmt.Errorf("receive request: %w", err)
 		}
-		for i := 0; i < messagesPerRequest; i++ {
-			if err := stream.Send(&pingv1.CumSumResponse{Sum: request.Number}); err != nil {
-				return fmt.Errorf("send response: %w", err)
-			}
+		if err := stream.Send(&pingv1.PingStreamResponse{
+			Id:   msg.Id,
+			Data: msg.Data,
+		}); err != nil {
+			return fmt.Errorf("send response: %w", err)
 		}
 	}
 }
 
-func cumSumFail(
+func pingStreamFail(
 	ctx context.Context,
-	stream *connect.BidiStream[pingv1.CumSumRequest, pingv1.CumSumResponse],
+	stream *connect.BidiStream[pingv1.PingStreamRequest, pingv1.PingStreamResponse],
 ) error {
 	if err := ctx.Err(); err != nil {
 		return err
 	}
-	request, err := stream.Receive()
+	_, err := stream.Receive()
 	if err != nil && errors.Is(err, io.EOF) {
 		return nil
 	}
-	if err := stream.Send(&pingv1.CumSumResponse{Sum: request.Number}); err != nil {
-		return fmt.Errorf("send response: %w", err)
-	}
-	if err := stream.Send(&pingv1.CumSumResponse{Sum: request.Number}); err != nil {
-		return fmt.Errorf("send response: %w", err)
-	}
 	return connect.NewError(connect.CodeDataLoss, errors.New("Oh no"))
 }
 
-func happyPingServer() *pluggablePingServer {
+func okayPingServer() *pluggablePingServer {
 	return &pluggablePingServer{
-		ping:   pingHappy,
-		cumSum: cumSumHappy,
+		ping:       pingOkay,
+		pingStream: pingStreamOkay,
 	}
 }
 
 func failPingServer() *pluggablePingServer {
 	return &pluggablePingServer{
-		ping:   pingFail,
-		cumSum: cumSumFail,
+		ping:       pingFail,
+		pingStream: pingStreamFail,
```

Comment:
If the two versions of this service (okay and fail) have different implementations of both methods, and we're only adding Etag headers on success, do we need this pluggable ping server abstraction at all? Could we just have two concrete types implementing the generated interface?

Reply:
There's two tests that use it. okay/fail service used to reduce boilerplate but we can probably still reduce.